CWatch Rewrite

JReissmueller opened the pull request:
Re-wrote the CWatch API.
Restructured services to create one account per service with licenses according to config options.
Modified the data shown in the site table and in the service expansion box.
Added input validation.
Added more logging.
Added the ability to remove a site from a license.

Commit:
Made numerous updates to formatting and language keys.
Made outputs more error safe.
Removed cwatch_sandbox for package fields.

Part of issue #6

Commit:
More formatting, error checking, and label changes.

Part of issue #6

Commit:
Started rewriting cwatch module

Commit:
Updated service tabs and service info partials.

Commit:
Updated API methods and error checking.
Abstracted view loading.
Updated/removed language.
Fixed formatting.

Commit:
Added more logging

Commit:
Abstracted code.
Added new templates for the add site form.
Updated service edits to give errors when reducing the number of licenses.
Updated README to describe necessary config options.

Commit:
Added service label.
Fixed api customer fetching and deletion.
Added tooltips for dns and ssl options.
Started adding action column on the sites tab.

Commit:
Added site removal.
Updated site info fetching
Fixed malware scanner adding and fetching.
Started moving malware control to a subsection of site management.

Commit:
Updated malware control to only allow registered domains.
Updated site list to show pending site provisions.

Commit:
Added input validation.

Commit:
Removed unused language

Commit:
Updated version to 2.0

Commit:
Added human readable site status.
Updated license label on service expand.
Malware scanner form is now populated with previously saved ftp info.

Commit:
Fixed site list to only include the proper sites/site provision requests.

Commit:
Fixed tab indentation

tysonphillips commented on api/cwatch_api.php:
```diff
@@ -0,0 +1,359 @@
+class CwatchApi
+{
+    // API endpoint URL
+    private $API_URL;
```
I think it would be more consistent to have this property be lower case

tysonphillips commented on api/cwatch_api.php:
```diff
@@ -0,0 +1,359 @@
+    public function __construct($username, $password, $sandbox)
+    {
+        if ($sandbox) {
+            $this->API_URL = 'http://cwatchpartnerportalstaging-env.us-east-1.elasticbeanstalk.com';
```
Does the sandbox endpoint not operate over https? We should always be using https, especially when making requests with credentials like a password.

JReissmueller replied:
The login at that url at least is inaccessible via https.

tysonphillips commented on api/cwatch_api.php:
```diff
@@ -0,0 +1,359 @@
+        $params = json_encode(['username' => $username, 'password' => $password]);
+        $response = $this->apiRequest('login', $params, 'POST');
+        if ($response) {
+            $auth = $response['headers'];
```
Assigning the response headers to an $auth variable seems to be redundant in this context. It would also be preferred to have this action be a separate method, e.g. setToken()

tysonphillips commented on api/cwatch_api.php:
```diff
@@ -0,0 +1,359 @@
+     * @param string $country The 3-character country code of the customer
+     * @return CwatchResponse
+     */
+    public function addUser($email, $firstName, $lastName, $country)
```
The variable formatting seems to be inconsistent with other variables--either they should follow under_score or camelCase format but not both

JReissmueller replied:
Fixed

tysonphillips commented on api/cwatch_api.php:
```diff
@@ -0,0 +1,359 @@
+        $params = json_encode(['email' => $email, 'name' => $firstName, 'surname' => $lastName, 'country' => $country]);
+        $response = $this->apiRequest('customer/add', $params, 'POST');
+
+        return new CwatchResponse($response['content']);
```
Instead of this, and every other method, returning a new CwatchResponse object, it would probably be better if the response from ::apiRequest() was a CwatchResponse object instead

JReissmueller replied:
That would be great, but the constructor uses the $response['header']

tysonphillips commented on api/cwatch_api.php:
```diff
@@ -0,0 +1,359 @@
+     * @param string $country The 3-character country code of the customer
+     * @return CwatchResponse
+     */
+    public function addLicense($licenseType, $term, $email, $firstName, $lastName, $country)
```
Same thing here with the camelCase vs under_score variable names

tysonphillips commented on api/cwatch_api.php:
```diff
@@ -0,0 +1,359 @@
+     *     - autoSsl Whether to install an ssl certificate
+     * @return CwatchResponse
+     */
+    public function addSite($params)
```
If this should be an array I would think the array keyword should be used to ensure it is of the correct type

tysonphillips commented on api/cwatch_api.php:
```diff
@@ -0,0 +1,359 @@
+     */
+    public function getSite($email, $domain)
+    {
+        $response = $this->apiRequest('customer/site/listBySite?email=' . $email . '&siteName=' . $domain, '', 'GET');
```
I notice that several requests are made to a URI followed by query arguments. It would probably be better to build this using http_build_query inside of ::apiRequest()

e.g.
```
class CwatchApi
{
    const GET = 'GET';
    const POST = 'POST';
    const PUT = 'PUT';
    const DELETE = 'DELETE';
...
public function getSite($email, $domain)
{
    ...
    return $this->apiRequest(self::GET, 'customer/site/listBySite', ['email' => $email, 'siteName' => $domain]);
}
...
private function apiRequest($method, $uri, array $data)
{
    // Set the data as query arguments in the GET request, otherwise it is body content
    if ($method === self::GET) {
        $uri .= !empty($data) ? '?' . http_build_query($data) : '';
    }

    ...
}
```

tysonphillips commented on api/cwatch_api.php:
```diff
@@ -0,0 +1,359 @@
+     */
+    public function getScanner($email, $domain)
+    {
+        $domainId = '';
```
Another under_score vs camelCase variable inconsistency

tysonphillips commented on api/cwatch_api.php:
```diff
@@ -0,0 +1,359 @@
+    {
+        $domainId = '';
+        $site_response = $this->getSite($email, $domain, '', 'GET');
+        if (empty($site_response->errorMsg)) {
```
So you check whether the response does not have an _errorMsg_ and then use _resp_. This seems like it would not necessarily work out well.

For instance, if the API cannot connect because the service is down, the connection will timeout and there will be no _errorMsg_, so it will try to use _resp_, which also does not exist, encountering an error.

Does the API have a better way to determine success? Return a 200 OK HTTP code we can check? Can the response be checked to be given before attempting to decode it?

tysonphillips commented on api/cwatch_api.php:
```diff
@@ -0,0 +1,359 @@
+    {
+        $domainId = '';
+        $site_response = $this->getSite($email, $domain, '', 'GET');
+        if (empty($site_response->errorMsg)) {
```
Same comments as ::getScanner

tysonphillips commented on api/cwatch_api.php:
```diff
@@ -0,0 +1,359 @@
+    public function addScanner($email, $params)
+    {
+        $domainId = '';
+        $site_response = $this->getSite($email, $params['domain'], '', 'GET');
```
Same comments as ::getScanner

tysonphillips commented on api/cwatch_api.php:
```diff
@@ -0,0 +1,359 @@
+
+        $result = curl_exec($ch);
+        if (curl_errno($ch)) {
+            throw new Exception('Curl Error: ' . curl_error($ch));
```
You're throwing an exception here but it's not caught anywhere in this source. Is it assumed this will be handled by the caller? If so, all public methods that call this should indicate they may @throw Exception

tysonphillips commented on cwatch.php:
```diff
@@ -9,16 +9,19 @@
  * @link http://blesta.com/ Blesta
  */
-class cwatch extends Module 
+class cwatch extends Module
```
Class names should be capitalized and follow CamelCase format, i.e. "Cwatch"

tysonphillips commented on cwatch.php:
```diff
@@ -252,57 +182,119 @@ public function getEmailTags()
+            $response = null;
+            try {
+                // Log request data
+                $this->log('createcommand', serialize($vars), 'input', true);
```
Ideally, we'd like to log only information that we're sending in the API request rather than all data submitted in $vars, especially if there is sensitive info (e.g. password) that should not be logged.

tysonphillips replied:
What would probably help out the most is for the CwatchApi class to have a method to retrieve the last request

```
private $last_request = [];

private function apiRequest(...) 
{
    $this->lastRequest = ['content' => ..., 'headers' => ''];
}
public function lastRequest()
{
    return $this->last_request;
}
```

tysonphillips commented on cwatch.php:
```diff
@@ -252,57 +182,119 @@ public function getEmailTags()
+                );
+
+                // Log response
+                $this->log('adduser', serialize($response), 'output', $response->code == 200);
```
The response is a CwatchResponse object. I would think we would log the raw response from the API only

tysonphillips commented on cwatch.php:
```diff
@@ -252,57 +182,119 @@ public function getEmailTags()
+                    }
+
+                    // Determine what term to add these licenses for
+                    $license_term = 'MONTH_1';
```
Shouldn't this fail if we don't have a pricing term that matches a valid license term? If I select a 3 month pricing period/term, why should the module add it for 1 month instead?

JReissmueller replied:
This is just a default.  It gets overridden when we fetch the pricing even if that pricing is not valid for cWatch.  So in your example this would be overridden with 'MONTH_3' and cWatch would return an error.

tysonphillips commented on cwatch.php:
```diff
@@ -252,57 +182,119 @@ public function getEmailTags()
+
+                            $this->log(
+                                'addlicense',
+                                serialize($license_response),
```
Same as my last comment on logging the raw data versus the CwatchResponse object. This should be applied to all places we do logging, and any expected sensitive information should be redacted in the log

tysonphillips commented on cwatch.php:
```diff
@@ -252,57 +182,119 @@ public function getEmailTags()
+
+        if (!empty($errors)) {
+            // Delete the user if something went wrong
+            $this->deleteUser(isset($vars['cwatch_email']) ? $vars['cwatch_email'] : '');
```
Why would you make an API call to delete a user represented by an empty string?

tysonphillips commented on cwatch.php:
```diff
@@ -328,75 +320,125 @@ public function addService($package, array $vars = null, $parent_package = null,
+                : $service_fields->cwatch_country;
+
+        if (isset($vars['use_module']) && $vars['use_module'] == 'true') {
+            try {
```
It looks like this does a bunch of stuff, which would be good for a separate method

```
if (isset($vars['use_module']) && $vars['use_module'] == 'true') {
    $fields = $this->updateService(...);
}

return $fields;
```

tysonphillips commented on api/cwatch_response.php:
```diff
@@ -1,17 +1,17 @@
 <?php
 
-class ApiResponse 
+class CwatchResponse
```
This class would be much better if it did not use public properties and used public methods instead.

```
public function status(); // whether the response was successful or not. might be http code, boolean, or something
public function raw(); // return the raw API response
public function response(); // converts the API response to object (json_decode, and on failure return a blank object)
public function errors(); // retrieves errors, if any
```

tysonphillips commented on cwatch.php:
```diff
@@ -417,37 +459,130 @@ public function suspendService($package, $service, $parent_package = null, $pare
-        }
-        $this->loadApi($user, $pass, $sandbox);
+        // Delete user
+        $this->deleteUser($service_fields->cwatch_email);
```
Is this cwatch_email field guaranteed to exist as a service field?

JReissmueller replied:
Depends on what you mean I guess.  The validation rules require it when adding a service, so yea I'd say so.

tysonphillips commented on cwatch.php:
```diff
@@ -460,106 +595,252 @@ public function cancelService($package, $service, $parent_package = null, $paren
-            'tabClientActions' => Language::_('Cwatch.tab_client_actions', true),
-            'tabClientMalWare' => Language::_('Cwatch.site.malware', true)
+            'tabClientSites' => Language::_('CWatch.tab_sites.sites', true),
+            'tabClientMalWare' => Language::_('CWatch.tab_malware.malware', true)
```
Is there a reason Malware is spelled with a capital W?

tysonphillips commented on cwatch.php:
```diff
@@ -460,106 +595,252 @@ public function cancelService($package, $service, $parent_package = null, $paren
+    {
+        return [
+            'tabSites' => Language::_('CWatch.tab_sites.sites', true),
+            'tabMalWare' => Language::_('CWatch.tab_malware.malware', true)
```
Is there a reason Malware is spelled with a capital W?

tysonphillips commented on cwatch.php:
```diff
@@ -460,106 +595,252 @@ public function cancelService($package, $service, $parent_package = null, $paren
      * @return string The string representing the contents of this tab
      */
-    function tabClientMalWare($package, $service, array $get = null, array $post = null, array $files = null) 
+    public function tabClientMalWare($package, $service, array $get = null, array $post = null, array $files = null)
```
Is there a reason Malware is spelled with a capital W?

tysonphillips commented on cwatch.php:
```diff
@@ -460,106 +595,252 @@ public function cancelService($package, $service, $parent_package = null, $paren
+     * @param array $files Any FILES parameters
+     * @return string The string representing the contents of this tab
+     */
+    public function tabMalWare($package, $service, array $get = null, array $post = null, array $files = null)
```
Is there a reason Malware is spelled with a capital W?

tysonphillips commented on cwatch.php:
```diff
@@ -617,40 +867,184 @@ public function getClientServiceInfo($service, $package)
+    /**
+     * Loads the cWatch API based on current row data
+     *
+     * @return \CwatchApi
```
You could always namespace the API so it doesn't belong to the global namespace

tysonphillips commented on cwatch.php:
```diff
@@ -617,40 +867,184 @@ public function getClientServiceInfo($service, $package)
+     * @param string $email The email name to validate
+     * @return bool True if the email is valid, false otherwise
+     */
+    public function validateEmail($email)
```
Input rule validation methods should just be anonymous functions

tysonphillips commented on views/default/client_add_site.pdt:
```diff
@@ -0,0 +1,40 @@
+</button>
+<?php
+$this->Form->end();
+?>
```
No blank new line? Colud omit the closing php tag

tysonphillips commented on views/default/manage.pdt:
```diff
@@ -1,69 +1,65 @@
+            $this->Form->fieldHidden('id', $this->Html->ifSet($module->id));
+            $this->Form->fieldHidden('row_id', $this->Html->ifSet($module->rows[$i]->id));
+            ?>
+            <a href="<?php echo $this->Html->safe($this->base_uri . 'settings/company/modules/deleterow/' . $this->Html->ifSet($module->id) . '/' . $this->Html->ifSet($module->rows[$i]->id) . '/'); ?>" class="manage" rel="<?php echo $this->Html->safe($this->_('CWatch.manage.confirm_delete', true)) . $module->rows[$i]->meta->username; ?>"><?php $this->_('CWatch.manage.delete'); ?></a>
```
The rel= attribute makes the language definition HTML-safe but not the username. We usually assume that language definitions are HTML-safe, but I would expect at least the username to be output HTML-safe

tysonphillips commented on views/default/tab_client_sites.pdt:
```diff
@@ -0,0 +1,62 @@
+        <td><?php $this->_('CWatch.tab_sites.actions'); ?></td>
+    </tr>
+    <?php
+    if (!empty($this->Html->ifSet($sites))) {
```
php's empty() construct cannot accept anything except variables < php 5.5. This should work on php 5.4 so the Html::ifSet call can be omitted.

tysonphillips commented on views/default/tab_sites.pdt:
```diff
@@ -0,0 +1,64 @@
+        <td><?php $this->_('CWatch.tab_sites.actions'); ?></td>
+    </tr>
+    <?php
+    if (!empty($this->Html->ifSet($sites))) {
```
php's empty() construct cannot accept anything except variables < php 5.5. This should work on php 5.4 so the Html::ifSet call can be omitted.

tysonphillips commented on views/default/manage.pdt:
```diff
@@ -1,69 +1,65 @@
-	if ($num_rows > 0) {
-	?>
+<?php
+$num_rows = count($this->Html->ifSet($module->rows));
```
This should default to an empty array rather than an empty string or it will not work on php 7.2+

Commit:
Updated formatting.
Made error checking more robust.

Commit:
Rewrote the CwatchResponse class.
Changed the API to handle errors better.
Changed the API to handle get parameters better.
Changed the API to handle response formatting better.
Updated logging.
Fixed spelling and formatting.
Abstracter code.

tysonphillips commented on api/cwatch_response.php:
```diff
@@ -0,0 +1,74 @@
+     * CwatchResponse constructor.
+     * @param array $api_response
+     */
+    public function __construct($api_response)
```
I would think the argument should include it's array type, and should work without error even if there is no 'content' or 'headers' keys.

tysonphillips commented on api/cwatch_response.php:
```diff
@@ -0,0 +1,74 @@
+    /**
+     * Get the status of this response
+     */
+    public function status()
```
Each of these methods should have an @return annotation.

tysonphillips commented on api/cwatch_api.php:
```diff
@@ -156,15 +144,13 @@ public function addSite($params)
-        if (empty($siteResponse->errorMsg)) {
-            $site = json_decode($siteResponse->resp);
+        $siteResponse = $this->getSite($email, $domain);
+        if (empty($siteResponse->errors())) {
```
"empty()" can only accept variables, not method calls, in php 5.4

tysonphillips commented on api/cwatch_api.php:
```diff
@@ -243,15 +231,13 @@ public function getLicenses($email)
-        if (empty($siteResponse->errorMsg)) {
-            $site = json_decode($siteResponse->resp);
+        $siteResponse = $this->getSite($email, $domain);
+        if (empty($siteResponse->errors())) {
```
"empty()" can only accept variables, not method calls, in php 5.4

tysonphillips commented on api/cwatch_api.php:
```diff
@@ -264,20 +250,23 @@ public function getScanner($email, $domain)
-        if (empty($siteResponse->errorMsg)) {
-            $site = json_decode($siteResponse->resp);
+        $siteResponse = $this->getSite($email, $domain);
+        if (empty($siteResponse->errors())) {
```
"empty()" can only accept variables, not method calls, in php 5.4

tysonphillips commented on api/cwatch_api.php:
```diff
@@ -287,65 +276,65 @@ public function getMalware($email, $domain)
-        if (empty($siteResponse->errorMsg)) {
-            $site = json_decode($siteResponse->resp);
+        $siteResponse = $this->getSite($email, $params['domain']);
+        if (empty($siteResponse->errors())) {
```
"empty()" can only accept variables, not method calls, in php 5.4

tysonphillips commented on api/cwatch_api.php:
```diff
@@ -287,65 +276,65 @@ public function getMalware($email, $domain)
-
         if ($method == 'POST') {
-            curl_setopt($ch, CURLOPT_POSTFIELDS, $body);
+            curl_setopt($ch, CURLOPT_POSTFIELDS, json_encode($body));
```
Couldn't the checks on $method be simplified?

e.g.
```
switch (strtoupper($method)) {
    default:
        curl_setopt($ch, CURLOPT_POSTFIELDS, json_encode($body));
        break;
    case 'GET':
        $url .= empty($body) ? '' : '?' . http_build_query($body);
        break;
}

curl_setopt($ch, CURLOPT_CUSTOMREQUEST, strtoupper($method));
````

tysonphillips commented on api/cwatch_api.php:
```diff
@@ -287,65 +276,65 @@ public function getMalware($email, $domain)
         }
 
         if ($method == 'GET') {
+            $url .= empty($body) ? '' : '?' . urldecode(http_build_query($body));
```
I think you can pass additional arguments to http_build_query so you don't need to call urldecode.

e.g.
`http_build_query($body, '', '&',  PHP_QUERY_RFC3986);`

JReissmueller replied:
Hmm, that just changes the encoding rules, it doesn't stop the encoding.

tysonphillips commented on api/cwatch_api.php:
```diff
@@ -287,65 +276,65 @@ public function getMalware($email, $domain)
         $result = curl_exec($ch);
         if (curl_errno($ch)) {
-            throw new Exception('Curl Error: ' . curl_error($ch));
+            return new CwatchResponse('{"error": "Curl Error", "message": "' . curl_error($ch) . '", "status": 500}');
```
```
$error = ['error' => 'Curl Error', 'message' => curl_error($ch), 'status' => 500];
return new CwatchResponse(json_encode($error));
```

Do we really want to show the cURL error though, or our own generic error instead? If users can see this in the interface it may not be good to display the curl error

tysonphillips commented on cwatch.php:
```diff
@@ -685,29 +705,31 @@ private function getMalwareTab($template, $service, $post)
             );
-            $this->log('addmalwarescanner', serialize($scanner), 'input', empty($scanner->errorMsg));
+            $last_request = $api->lastRequest();
+            $last_request['password'] = isset($last_request) ? '***' : '';
```
I thought the last request only contains an array of 'content' and 'headers' keys? You're adding a 'password' key but not hiding the password in the content?

tysonphillips commented on cwatch.php:
```diff
@@ -685,29 +705,31 @@ private function getMalwareTab($template, $service, $post)
 
-            if (!empty($scanner->errorMsg)) {
-                $this->Input->setErrors(['api' => ['internal' => $scanner->errorMsg]]);
+            if (!empty($scanner->errors())) {
```
"empty()" can only accept variables, not method calls, in php 5.4

tysonphillips commented on cwatch.php:
```diff
@@ -685,29 +705,31 @@ private function getMalwareTab($template, $service, $post)
         $domains_ftp = [];
-        if (empty($sites_response->errorMsg)) {
-            foreach (json_decode($sites_response->resp) as $site) {
+        if (empty($sites_response->errors())) {
```
"empty()" can only accept variables, not method calls, in php 5.4

tysonphillips commented on cwatch.php:
```diff
@@ -685,29 +705,31 @@ private function getMalwareTab($template, $service, $post)
                 $scanner_response = $api->getScanner($service_fields->cwatch_email, $site->domain);
-                if (empty($scanner_response->errorMsg)) {
-                    $scanner = json_decode($scanner_response->resp);
+                if (empty($scanner_response->errors())) {
```
"empty()" can only accept variables, not method calls, in php 5.4

tysonphillips commented on cwatch.php:
```diff
@@ -785,9 +806,10 @@ private function getSitesTab($template, $service, $post)
-                if (!empty($site->errorMsg)) {
-                    $this->Input->setErrors(['api' => ['internal' => $site->errorMsg]]);
+                $this->log('addsite', serialize($api->lastRequest()), 'input', true);
+                $this->log('addsite', $site->raw(), 'output', empty($site->errors()));
```
"empty()" can only accept variables, not method calls, in php 5.4

tysonphillips commented on cwatch.php:
```diff
@@ -785,9 +806,10 @@ private function getSitesTab($template, $service, $post)
-                    $this->Input->setErrors(['api' => ['internal' => $site->errorMsg]]);
+                $this->log('addsite', serialize($api->lastRequest()), 'input', true);
+                $this->log('addsite', $site->raw(), 'output', empty($site->errors()));
+                if (!empty($site->errors())) {
```
"empty()" can only accept variables, not method calls, in php 5.4

tysonphillips commented on cwatch.php:
```diff
@@ -685,29 +705,31 @@ private function getMalwareTab($template, $service, $post)
+            $last_request = $api->lastRequest();
+            $last_request['password'] = isset($last_request) ? '***' : '';
+            $this->log('addmalwarescanner', serialize($last_request), 'input', true);
+            $this->log('addmalwarescanner', $scanner->raw(), 'output', empty($scanner->errors()));
```
"empty()" can only accept variables, not method calls, in php 5.4

tysonphillips commented on cwatch.php:
```diff
@@ -797,11 +819,11 @@ private function getSitesTab($template, $service, $post)
 
-        if (empty($sites_response->errorMsg)) {
-            $site_list = array_merge($site_list, json_decode($sites_response->resp));
+        if (empty($sites_response->errors())) {
```
"empty()" can only accept variables, not method calls, in php 5.4

tysonphillips commented on cwatch.php:
```diff
@@ -797,11 +819,11 @@ private function getSitesTab($template, $service, $post)
         }
-        if (empty($provisions_response->errorMsg)) {
-            $site_list = array_merge($site_list, json_decode($provisions_response->resp));
+        if (empty($provisions_response->errors())) {
```
"empty()" can only accept variables, not method calls, in php 5.4

tysonphillips commented on cwatch.php:
```diff
@@ -812,14 +834,14 @@ private function getSitesTab($template, $service, $post)
                 $scanner = $api->getScanner($service_fields->cwatch_email, $site->domain);
-                if (empty($scanner->errorMsg)) {
-                    $site->scanner = json_decode($scanner->resp);
+                if (empty($scanner->errors())) {
```
"empty()" can only accept variables, not method calls, in php 5.4

tysonphillips commented on cwatch.php:
```diff
@@ -812,14 +834,14 @@ private function getSitesTab($template, $service, $post)
                 $license = $api->getLicense($site->licenseKey);
-                if (empty($license->errorMsg)) {
-                    $site->license = json_decode($license->resp);
+                if (empty($license->errors())) {
```
"empty()" can only accept variables, not method calls, in php 5.4

tysonphillips commented on cwatch.php:
```diff
@@ -829,10 +851,10 @@ private function getSitesTab($template, $service, $post)
-            foreach (json_decode($licenses_response->resp, true) as $license) {
-                if (strtolower($license['status']) == 'valid' && $license['registeredDomainCount'] == 0) {
-                    $licenses[$license['licenseKey']] = $license['productTitle'];
+        if (empty($licenses_response->errors())) {
```
"empty()" can only accept variables, not method calls, in php 5.4

tysonphillips commented on cwatch.php:
```diff
@@ -894,8 +916,8 @@ private function getServiceInfo($template, $service)
         $licenses = [];
-        if (empty($licenses_response->errorMsg)) {
-            foreach (json_decode($licenses_response->resp) as $license) {
+        if (empty($licenses_response->errors())) {
```
"empty()" can only accept variables, not method calls, in php 5.4

tysonphillips commented on cwatch.php:
```diff
@@ -989,7 +1011,21 @@ private function getServiceRules(array $vars = null, $edit = false)
+                        $api = $this->getApi();
+                        $user_response = $api->getUser($email);
+
+                        if (empty($user_response->errors())) {
```
"empty()" can only accept variables, not method calls, in php 5.4

Commit:
Updated config file and readme to show correct name for 30 day trial license type

Commit:
Updated error checking to work with php 5.4.
Simplified code.
Updated comments.
Updated log filtering.

Commit:
Updated error checking to work with php 5.4.

Commit:
Site provisions are no longer allowed for licenses currently attempting a site provision.
